README.md
# DiscoDNC

Deterministic Network Calculus (DNC) is a methodology for worst-case modeling and analysis of communication networks. It enables to derive deterministic bounds on a server’s backlog as well as a flow’s end-to-end delay. Given a directed graph of servers (server graph) and the flows crossing these servers, the Disco Deterministic Network Calculator (DiscoDNC) automates the derivation of bounds.

# ECES Fork

This is a fork of the 2.4.3 version of the [Disco DNC project (now networkcalculus.org DNC project)]( https://github.com/NetCal/DNC).
A few minor modifications have been made in order to match the needs of the [*deterministic network modeling* (DNM)](https://github.com/AmoVanB/eces-dnm) module of the [ECES](https://github.com/AmoVanB/eces-core) framework developed at the [Chair of Communication Networks (LKN)](https://www.lkn.ei.tum.de) of the Technical University of Munich.
A comprehensive list of the changes can be found in the `Changelog.txt` file.


Changelog.txt
v2.4.3-lkn (2019-Apr-03)
------------------------

* Removal of the unused RTC toolbox
* Usage of the same maven-surefire-plugin version as for our other modules
* Addition of two token buckets leads to a new token bucket (and not "just" an arrival curve)
* Addition of a token bucket and a constant leads to a new token bucket (and not "just" an arrival curve)
* Still consider a token bucket with null rate as a token bucket
* Enable the conversion of a service curve to a rate-latency curve if it indeed has such a shape
* Some tab -> spaces here and there

v2.4.3 "Chimera" (2018-Aug-03)
------------------------------
* Fixed Bug #31: Computation of infinite output bound due to stability constraint violation.
* Improved shifting of curves (horizontal and vertical).
* SFA tandem analysis: Shortcut in case there is no interference by other flows. 
* Comment cleanups and improvements.

v2.4.2 "Chimera" (2018-Jul-04)
------------------------------
* Tandem Matching Analysis and Arrival Bound added 
* SFA and PbooArrivalBounds:
  - Improved structure, reliability and performance.
  - Single, static SFA tandem analysis for SFA and PBOO ABs. 
* Rational Num implementations gained a getValue function returning a new instance of the Math3 object in their backend.
* Minor Javadocs and comment improvements.

v2.4.1 "Chimera" (2018-May-07)
------------------------------
* Tests
  - They now correctly consider the configures Curve and Num implementation (GitHub Issue #1).
  - Each test only works with a single instance of its respective network instead of creating one per test config.
  - We use external method sources to not forcibly change configs to comply with PMOO's restriction to arbitrary multiplexing.
  - Expected test results are were moved to separate classes. 
* Added a class with presets with most common flow of interest analysis and arrival bounding configurations
  for algebraic compositional feed-forward analysis (compFFA): de.uni_kl.cs.discodnc.nc.CompFFApresets. 
  - Demo4 uses the presets instead of analyses instances implicitly using the defailt configuration.
* Our current TFA AB burst reduction implementation is only applicable to TB arrivals.
  - We now skip reducing the reduction for arrival curves with >2 segments.
* Network: You can now get a flow by its ID.
* SeparateFlowAnalysis: Do not rely on an exception to identify path's source server.
* Improved pom.xml.
* Code and comment cleanups.
  - Rename location of interference, loi, to server of interference, soi, to prevent confusion with potential link of interference.

v2.4.0 "Chimera" (2018-Mar-18)
------------------------------
* Maven build system
  - Reworked the source folders to src/main/java and src/test/java.
  - Added pom.xml
  - Attention: You need to add the MPA toolbox manually:
     - download it RTCToolbox_bin.zip at http://www.mpa.ethz.ch/Rtctoolbox/Download
     - unzip it, and put the file RTCToolbox/rtc/rtc.jar into the DiscoDNC's lib folder.
  - Removed Ant files
* New domain discodnc.cs.uni-kl.de
  - Includes a new webpage.
  - Package prefixes have been renamed to de.uni_kl.cs.discondc.
* All the changes of betas 1 to 4.
  - Attention: Although all tests succeed with the rtc curve backend, it is considered a beta feature. 
* Static analysis update: switched from obsolete FindBugs to its successor SpotBugs.
  - Fixed EQ_CHECK_FOR_OPERAND_NOT_COMPATIBLE_WITH_THIS in LinearSegment_MPARTC_PwAffine.
  - Fixed RV_RETURN_VALUE_IGNORED_BAD_PRACTICE in Network.
* Reworked storing a network to the file system.
* Removed broken OutputTest.
* Code cleanups.
  - Moved preliminary code to dispatch to different DNC bounding implementations to comments.

v2.4.0beta4 "Chimera" (2017-Dec-23)
-----------------------------------
* Attention: The DiscoDNC does not yet fully work with Java 9. Compiler compliance level 1.8 is required.
  - Thanks to Zoltán Mátyás for pointing this out.
* Tests are all ported to JUnit5.
* Check the classpath for rtc.jar instead of using a hardcoded location (patch by Zoltán Mátyás). 
* DNC bounds derived from two curves (i.e., other than arrival bound algorithms)
  - package renamed from de.uni_kl.cs.disco.nc.operations to de.uni_kl.cs.disco.nc.bounds
  - Computations (Backlog, Delay, LeftOverService and Output) are called via a dispatcher, Bound.java.
  - Name simplifications: drop "Bound" suffix for Backlog, Delay, and Output classes.  
* Method cleanups in curve classes. Code improvements apply to the DNC implementation:
  - Renamed auto-generated method names that started with "isIs" or "isHas" and removed underscores from these method names.
  - Rate latencies: merged getRL_Property() into isRateLateny(); renamed setRL_Property() to setRateLateny().
  - Token buckets: merged getTB_Property() into isTokenBucket(); renamed setTB_Property() to setTokenBucket().
  - getDelayedInfiniteBurst_Property renamed to isDelayedInfiniteBurst.
  - Merged getTB_Burst into getBurst.
  - Created a rate-latency optimized code path for getLatenty(), similar to getBurst().
* Comments and JavaDocs cleanups, various stability and performance improvements.

v2.4.0beta3 "Chimera" (2017-Sep-01)
-----------------------------------
Simplifications to the code structure for streamlined use. This work requires Java 8+ as it relies on static methods in interfaces.
* CalculatorConfig singleton: fix execution of all tests, run time back to previous level.
* Split Analysis.java into three parts:
  - The Analysis interface that also provides static methods for simple execution of analyses.
  - The abstract class AbstractAnalysis holds member variables for the network, the configurations and the results.
  All analyses in de.uni_kl.cs.disco.nc.analyses extend AbstractAnalysis and implement Analysis.
 * Split ArrivalBound.java into three parts:
  - The minimal ArrivalBound interface.
  - The abstract class ArrivalBoundDispatch that applies the configured arrival boundings.

src/main/java/de/uni_kl/cs/discodnc/curves/LinearSegment.java
/*
 * This file is part of the Disco Deterministic Network Calculator.
 *
 * Copyright (C) 2017+ The DiscoDNC contributors
 *
 * disco | Distributed Computer Systems Lab
 * University of Kaiserslautern, Germany
 *
 * http://discodnc.cs.uni-kl.de
 *
 *
 * The Disco Deterministic Network Calculator (DiscoDNC) is free software;
 * you can redistribute it and/or modify it under the terms of the 
 * GNU Lesser General Public License as published by the Free Software Foundation; 
 * either version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

package de.uni_kl.cs.discodnc.curves;

import de.uni_kl.cs.discodnc.curves.dnc.LinearSegment_DNC;
import de.uni_kl.cs.discodnc.nc.CalculatorConfig;
import de.uni_kl.cs.discodnc.numbers.Num;

public interface LinearSegment {
    static LinearSegment createLinearSegment(Num x, Num y, Num grad, boolean leftopen) {
        switch (CalculatorConfig.getInstance().getCurveImpl()) {
            case MPA_RTC:
                throw new RuntimeException("We disable the usage of the RTC Toolbox!");
            case DNC:
            default:
                return new LinearSegment_DNC(x, y, grad, leftopen);
        }
    }

    static LinearSegment createHorizontalLine(double y) {
        switch (CalculatorConfig.getInstance().getCurveImpl()) {
            case MPA_RTC:
                throw new RuntimeException("We disable the usage of the RTC Toolbox!");
            case DNC:
            default:
                return new LinearSegment_DNC(Num.getFactory().createZero(),
                        Num.getFactory().create(y), Num.getFactory().createZero(), false);
        }
    }

    /**
     * Helper creating a new segment starting at x that is the sum of the given
     * getSegment.
     *
     * @param s1       Segment 1.
     * @param s2       Segment 2.
     * @param x        New x-coordinate the start at.
     * @param leftopen Set the segment to be left-open.
     * @return The new linear segment, pointwise sum of the given ones, starting in
     * x.
     */
    static LinearSegment add(LinearSegment s1, LinearSegment s2, Num x, boolean leftopen) {
        LinearSegment result = createHorizontalLine(0.0);
        result.setX(x);
        result.setY(Num.getUtils().add(s1.f(x), s2.f(x)));
        result.setGrad(Num.getUtils().add(s1.getGrad(), s2.getGrad()));
        result.setLeftopen(leftopen);
        return result;
    }

    /**
     * Helper creating a new segment starting at x that is the difference between
     * the given getSegment.
     *
     * @param s1       Segment 1.
     * @param s2       Segment 2.
     * @param x        New x-coordinate the start at.
     * @param leftopen Set the segment to be left-open.
     * @return The new linear segment, pointwise difference of the given ones, i.e.,
     * s1 - s2, starting in x.
     */
    static LinearSegment sub(LinearSegment s1, LinearSegment s2, Num x, boolean leftopen) {
        LinearSegment result = createHorizontalLine(0.0);
        result.setX(x);
        result.setY(Num.getUtils().sub(s1.f(x), s2.f(x)));
        result.setGrad(Num.getUtils().sub(s1.getGrad(), s2.getGrad()));
        result.setLeftopen(leftopen);
        return result;
    }

    /**
     * Helper creating a new segment starting at x that is the minimum of the given
     * getSegment. Note: Only valid if s1 and s2 do not intersect before the next
     * IP.
     *
     * @param s1       Segment 1.
     * @param s2       Segment 2.
     * @param x        New x-coordinate the start at.
     * @param leftopen Set the segment to be left-open.
     * @param crossed  Provides information if the segments intersect.
     * @return The new linear segment, pointwise minimum of the given ones, starting
     * in x.
     */
    static LinearSegment min(LinearSegment s1, LinearSegment s2, Num x, boolean leftopen, boolean crossed) {
        Num f1_x = s1.f(x);
        Num f2_x = s2.f(x);

        LinearSegment result = createHorizontalLine(0.0);
        result.setX(x);
        if (crossed || Num.getUtils().abs(Num.getUtils().sub(f1_x, f2_x))
                .lt(Num.getFactory().getEpsilon())) {
            result.setY(f1_x);
            result.setGrad(Num.getUtils().min(s1.getGrad(), s2.getGrad()));
        } else if (f1_x.lt(f2_x)) {
            result.setY(f1_x);
            result.setGrad(s1.getGrad());
        } else {
            result.setY(f2_x);
            result.setGrad(s2.getGrad());
        }
        result.setLeftopen(leftopen);
        return result;
    }

    /**
     * Helper creating a new segment starting at x that is the maximum of the given
     * segments. Note: Only valid if s1 and s2 do not intersect before the next IP.
     *
     * @param s1       Segment 1.
     * @param s2       Segment 2.
     * @param x        New x-coordinate the start at.
     * @param leftopen Set the segment to be left-open.
     * @param crossed  Provides information if the segments intersect.
     * @return The new linear segment, pointwise maximum of the given ones, starting
     * in x.
     */
    static LinearSegment max(LinearSegment s1, LinearSegment s2, Num x, boolean leftopen, boolean crossed) {
        Num f1_x = s1.f(x);
        Num f2_x = s2.f(x);

        LinearSegment result = createHorizontalLine(0.0);
        result.setX(x);
        if (crossed || Num.getUtils().abs(Num.getUtils().sub(f1_x, f2_x))
                .lt(Num.getFactory().getEpsilon())) {
            result.setY(f1_x);
            result.setGrad(Num.getUtils().max(s1.getGrad(), s2.getGrad()));
        } else if (f1_x.gt(f2_x)) {
            result.setY(f1_x);
            result.setGrad(s1.getGrad());
        } else {
            result.setY(f2_x);
            result.setGrad(s2.getGrad());
        }
        result.setLeftopen(leftopen);
        return result;
    }

    // --------------------------------------------------------------------------------------------------------------
    // Interface
    // --------------------------------------------------------------------------------------------------------------
    Num f(Num x);

    Num getX();

    void setX(Num x);

    Num getY();

    void setY(Num y);

    Num getGrad();

    void setGrad(Num grad);

    boolean isLeftopen();

    // --------------------------------------------------------------------------------------------------------------
    // Factory
    // --------------------------------------------------------------------------------------------------------------
    // In contrast to the Curve factory, it does not dispatch to other factories.

    void setLeftopen(boolean leftopen);

    Num getXIntersectionWith(LinearSegment other);

    // --------------------------------------------------------------------------------------------------------------
    // Utils
    // --------------------------------------------------------------------------------------------------------------

    LinearSegment copy();

    @Override
    boolean equals(Object obj);

    @Override
    int hashCode();

    @Override
    String toString();
}


src/main/java/de/uni_kl/cs/discodnc/curves/dnc/ServiceCurve_DNC.java
/*
 * This file is part of the Disco Deterministic Network Calculator.
 *
 * Copyright (C) 2005 - 2007 Frank A. Zdarsky
 * Copyright (C) 2013 - 2018 Steffen Bondorf
 * Copyright (C) 2017+ The DiscoDNC contributors
 *
 * disco | Distributed Computer Systems Lab
 * University of Kaiserslautern, Germany
 *
 * http://discodnc.cs.uni-kl.de
 *
 *
 * The Disco Deterministic Network Calculator (DiscoDNC) is free software;
 * you can redistribute it and/or modify it under the terms of the 
 * GNU Lesser General Public License as published by the Free Software Foundation; 
 * either version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

package de.uni_kl.cs.discodnc.curves.dnc;

import de.uni_kl.cs.discodnc.curves.CurvePwAffine;
import de.uni_kl.cs.discodnc.curves.ServiceCurve;
import de.uni_kl.cs.discodnc.nc.CalculatorConfig;
import de.uni_kl.cs.discodnc.numbers.Num;

public class ServiceCurve_DNC extends Curve_DNC implements ServiceCurve {
    // --------------------------------------------------------------------------------------------------------------
    // Constructors
    // --------------------------------------------------------------------------------------------------------------
    public ServiceCurve_DNC() {
        super();
    }

    public ServiceCurve_DNC(int segment_count) {
        super(segment_count);
    }

    public ServiceCurve_DNC(CurvePwAffine curve) {
        copy(curve);

        // Too strong requirement: !isConvex()
        if (CalculatorConfig.getInstance().exec_service_curve_checks() && !isWideSenseIncreasing()) {
            throw new RuntimeException("Service curves can only be created from wide-sense increasing functions.");
        }
    }

    public ServiceCurve_DNC(String service_curve_str) throws Exception {
        // Smallest possible string: {(0,0),0}
        if (service_curve_str == null || service_curve_str.isEmpty() || service_curve_str.length() < 9) {
            throw new RuntimeException("Invalid string representation of a service curve.");
        }

        initializeCurve(service_curve_str);

        // Too strong requirement: !isConvex()
        if (CalculatorConfig.getInstance().exec_service_curve_checks() && !isWideSenseIncreasing()) {
            throw new RuntimeException("Service curves can only be created from wide-sense increasing functions.");
        }
    }

    // --------------------------------------------------------------------------------------------------------------
    // Interface Implementations
    // --------------------------------------------------------------------------------------------------------------
    @Override
    public ServiceCurve_DNC copy() {
        ServiceCurve_DNC sc_copy = new ServiceCurve_DNC();
        sc_copy.copy(this);
        return sc_copy;
    }

    public void makeRateLatency() {
        if(this.segments.length == 2
                && this.segments[0].x.eq(0)
                && this.segments[0].y.eq(0)
                && this.segments[0].grad.eq(0)
                && this.segments[1].y.eq(0)
                && this.segments[1].x.geq(Num.getFactory().create(0))
                && this.segments[1].grad.geq(Num.getFactory().create(0))) {
            // Ok it is a rate-latency curve!
            makeRateLatency(this, this.segments[1].grad, this.segments[1].x);
        }
        else {
            throw new RuntimeException("This curve in not a rate latency curve!");
        }
    }

    @Override
    public boolean equals(Object obj) {
        return (obj instanceof ServiceCurve_DNC) && super.equals(obj);
    }

    @Override
    public int hashCode() {
        return "SC".hashCode() * super.hashCode();
    }

    /**
     * Returns a string representation of this curve.
     *
     * @return the curve represented as a string.
     */
    @Override
    public String toString() {
        return "SC" + super.toString();
    }
}


src/main/java/de/uni_kl/cs/discodnc/minplus/MinPlus.java
/*
 * This file is part of the Disco Deterministic Network Calculator.
 *
 * Copyright (C) 2017+ The DiscoDNC contributors
 *
 * disco | Distributed Computer Systems Lab
 * University of Kaiserslautern, Germany
 *
 * http://discodnc.cs.uni-kl.de
 *
 *
 * The Disco Deterministic Network Calculator (DiscoDNC) is free software;
 * you can redistribute it and/or modify it under the terms of the 
 * GNU Lesser General Public License as published by the Free Software Foundation; 
 * either version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

package de.uni_kl.cs.discodnc.minplus;

// Due to name collisions, these classes are not imported,
// they are referenced by their fully qualified names.
//import ch.ethz.rtc.kernel.Curve;
//import Curve;


import de.uni_kl.cs.discodnc.curves.ArrivalCurve;
import de.uni_kl.cs.discodnc.curves.Curve;
import de.uni_kl.cs.discodnc.curves.CurvePwAffine;
import de.uni_kl.cs.discodnc.curves.MaxServiceCurve;
import de.uni_kl.cs.discodnc.curves.ServiceCurve;
import de.uni_kl.cs.discodnc.minplus.dnc.Convolution_DNC;
import de.uni_kl.cs.discodnc.minplus.dnc.Deconvolution_DNC;
import de.uni_kl.cs.discodnc.nc.CalculatorConfig;
import de.uni_kl.cs.discodnc.nc.CalculatorConfig.CurveImpl;
import de.uni_kl.cs.discodnc.nc.CalculatorConfig.OperationImpl;

import java.util.HashSet;
import java.util.Set;

public abstract class MinPlus {
	// --------------------------------------------------------------------------------------------------------------
	// Min-Plus-Operation Dispatching
	// --------------------------------------------------------------------------------------------------------------

	// ------------------------------------------------------------
	// Convolution
	// ------------------------------------------------------------

	// Service Curves
	public static ServiceCurve convolve(ServiceCurve service_curve_1, ServiceCurve service_curve_2) throws Exception {
		return convolve(service_curve_1, service_curve_2, false);
	}

	public static ServiceCurve convolve(ServiceCurve service_curve_1, ServiceCurve service_curve_2,
			boolean tb_rl_optimized) throws Exception {
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC) // DNC operations work with
				// DNC and MPA_RTC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) { // NATIVE operation on
			// DNC curves
			return Convolution_DNC.convolve(service_curve_1, service_curve_2, tb_rl_optimized);

		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	// Java won't let us call this method "convolve" because it does not care about
	// the Sets' types; tells that there's already another method taking the same
	// arguments.
	public static Set<ServiceCurve> convolve_SCs_SCs(Set<ServiceCurve> service_curves_1,
			Set<ServiceCurve> service_curves_2) throws Exception {
		return convolve_SCs_SCs(service_curves_1, service_curves_2, false);
	}

	public static Set<ServiceCurve> convolve_SCs_SCs(Set<ServiceCurve> service_curves_1,
			Set<ServiceCurve> service_curves_2, boolean tb_rl_optimized) throws Exception {

		if (service_curves_1.isEmpty()) {
			return service_curves_2;
		}
		if (service_curves_2.isEmpty()) {
			return service_curves_1;
		}

		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC) // DNC operations work with
				// DNC and MPA_RTC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) { // NATIVE operation on
			// DNC curves
			return Convolution_DNC.convolve_SCs_SCs(service_curves_1, service_curves_2, tb_rl_optimized);

		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	// Arrival Curves
	public static ArrivalCurve convolve(ArrivalCurve arrival_curve_1, ArrivalCurve arrival_curve_2) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Convolution_DNC.convolve(arrival_curve_1, arrival_curve_2);

		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	public static ArrivalCurve convolve(Set<ArrivalCurve> arrival_curves) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Convolution_DNC.convolve(arrival_curves);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	// Maximum Service Curves
	public static MaxServiceCurve convolve(MaxServiceCurve max_service_curve_1, MaxServiceCurve max_service_curve_2)
			throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Convolution_DNC.convolve(max_service_curve_1, max_service_curve_2);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	// Arrival Curves and Max Service Curves
	public static Set<CurvePwAffine> convolve_ACs_MSC(Set<ArrivalCurve> arrival_curves,
			MaxServiceCurve maximum_service_curve) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Convolution_DNC.convolve_ACs_MSC(arrival_curves, maximum_service_curve);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	public static Set<ArrivalCurve> convolve_ACs_EGamma(Set<ArrivalCurve> arrival_curves,
			MaxServiceCurve extra_gamma_curve) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Convolution_DNC.convolve_ACs_EGamma(arrival_curves, extra_gamma_curve);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	// ------------------------------------------------------------
	// Deconvolution
	// ------------------------------------------------------------
	public static Set<ArrivalCurve> deconvolve(Set<ArrivalCurve> arrival_curves, ServiceCurve service_curve)
			throws Exception {
		return deconvolve(arrival_curves, service_curve, false);
	}

	public static Set<ArrivalCurve> deconvolve(Set<ArrivalCurve> arrival_curves, ServiceCurve service_curve,
			boolean tb_rl_optimized) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Deconvolution_DNC.deconvolve(arrival_curves, service_curve, tb_rl_optimized);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	public static Set<ArrivalCurve> deconvolve(Set<ArrivalCurve> arrival_curves, Set<ServiceCurve> service_curves)
			throws Exception {
		return deconvolve(arrival_curves, service_curves, false);
	}

	public static Set<ArrivalCurve> deconvolve(Set<ArrivalCurve> arrival_curves, Set<ServiceCurve> service_curves,
			boolean tb_rl_optimized) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Deconvolution_DNC.deconvolve(arrival_curves, service_curves, tb_rl_optimized);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	public static ArrivalCurve deconvolve(ArrivalCurve arrival_curve, ServiceCurve service_curve) throws Exception {
		return deconvolve(arrival_curve, service_curve, false);
	}

	public static ArrivalCurve deconvolve(ArrivalCurve arrival_curve, ServiceCurve service_curve,
			boolean tb_rl_optimized) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Deconvolution_DNC.deconvolve(arrival_curve, service_curve);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	public static Set<ArrivalCurve> deconvolve_almostConcCs_SCs(Set<CurvePwAffine> curves,
			Set<ServiceCurve> service_curves) throws Exception {
		// DNC operations work with DNC and MPA_RTC curves
		if (CalculatorConfig.getInstance().getOperationImpl().equals(OperationImpl.DNC)
				// NATIVE operation on DNC curves
				|| CalculatorConfig.getInstance().getCurveImpl().equals(CurveImpl.DNC)) {
			return Deconvolution_DNC.deconvolve_almostConcCs_SCs(curves, service_curves);
		} else { // Must be CurveClass.MPA_RTC + OpertionClass.NATIVE
			throw new RuntimeException("We disable the usage of the RTC Toolbox!");
		}
	}

	// --------------------------------------------------------------------------------------------------------------
	// Min-Plus-Operation Input Checks
	// --------------------------------------------------------------------------------------------------------------

	/**
	 * @param obj1
	 * @param obj2
	 * @return 0 == none of the objects is null,
	 *         1 == the first object is null,
	 *         2 == the second object is null,
	 *         3 == both objects are null.
	 */
	public static int inputNullCheck(Object obj1, Object obj2) {
		// Usually neither is null so this initial check promises best overall
		// performance.
		if (obj1 != null && obj2 != null) {
			return 0;
		}

		int return_value = 0;

		if (obj1 == null) {
			return_value += 1;
		}
		if (obj2 == null) {
			return_value += 2;
		}

		return return_value;
	}

	/**
	 * @param curve_1
	 * @param curve_2
	 * @return 0 == none of the objects is a delayed infinite burst,
	 *         1 == the first object is a delayed infinite burst,
	 *         2 == the second object is a delayed infinite burst,
	 *         3 == both objects are a delayed infinite burst.
	 */
	public static int inputDelayedInfiniteBurstCheck(Curve curve_1, Curve curve_2) {
		int return_value = 0;

		if (curve_1.isDelayedInfiniteBurst()) {
			return_value += 1;
		}

		if (curve_2.isDelayedInfiniteBurst()) {
			return_value += 2;
		}

		return return_value;
	}

	/**
	 * @param set1
	 * @param set2
	 * @return 0 == none of the sets is empty,
	 *         1 == the first sets is empty,
	 *         2 == the second sets is empty,
	 *         3 == both sets are empty.
	 */
	@SuppressWarnings("rawtypes")
	public static int inputEmptySetCheck(Set set1, Set set2) {
		// Usually neither is empty so this initial check promises best overall
		// performance.
		if (!set1.isEmpty() && !set2.isEmpty()) {
			return 0;
		}

		int return_value = 0;

		if (set1.isEmpty()) {
			return_value += 1;
		}
		if (set2.isEmpty()) {
			return_value += 2;
		}

		return return_value;
	}
}
